### ISSUE TITLE
.,kmj,

### ISSUE THREAD

**Issue by lissom** [restic/restic#5135]
Output of `restic version`
--------------------------
restic 0.16.3 compiled with go1.21.6 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
backup --tags-success
Tags that only apply if the exit code is zero.
We want all backups, but we also need to know which ones are complete vs exit code 3. 
I can be very difficult to tell that files are missing until something larger goes wrong down the road.

(Storing if the errors in the snapshot data could work too.)

What are you trying to do? What problem would this solve?
---------------------------------------------------------
Detect good vs bad backups.  We want all of the backup, good or not.  So ignoring the snapshot on failure isn't an ideal solution, though this would be preferable to the current situation.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Restic is great.

**Comment by MichaelEischer** [comment]
An alternative would be to extend the snapshot summary [`SnapshotSummary`](https://restic.readthedocs.io/en/stable/075_scripting.html#snapshots). That information is available for example via `snapshots --json`. I somewhat prefer that variant as can always be enabled and doesn't require yet another option for the backup command.

> (Storing if there are errors in the snapshot data could work too. Something as granular as the files that couldn't be read would be great.)

Storing the backup log would require a whole separate mechanism as logs can be quite big.

**Comment by lissom** [comment]
That would work too. However, unless it's searchable (like a tag) then I would still need to tag it to get the correct snapshot in the usual way.

**Comment by Gaibhne** [comment]
I agree, unless metadata gets made searchable/filterable through the CLI, not adding a regular command line flag for it would mean reading back and manually/via script parsing all existing snapshots' metadata, which does not seem like a reliable/dependable path for something as critical as handling/dealing with incomplete snapshots, which essentially imply lost or potentially lost data.

I would argue that data integrity is the absolute number one focus of `restic` (and backup tools in general), not a nice convenience feature that can be hidden behind user scripts shared on forum posts. As such, I would definitely think it reasonable to extend the backup options.

My suggestion would be an opt-out solution that tags incomplete snapshots. Incomplete snapshots are the exception, and people noticing the tag by surprise in their repo would be a good thing. In that spirit, a `--no-tag-unsuccessful` or `--no-tag-with-errors` or similar would be my ideal way forward.

**Comment by lissom** [comment]
#5362 would be great, even better if it can be searched for like tags are.

**Comment by rawtaz** [comment]
I very much agree with @MichaelEischer that the information needed for finding the snapshots you're looking for, in this case for example the exit code, should be put in the snapshot metadata instead of in tags. Putting it in tags opens up a few other points of discussion and is just messy, compared to having it in the metadata in a well structured way.

If we assume for example that there's a field `exit_code` in the snapshots' metadata, one could easily find snapshots with e.g. exit code 3 like this: `restic snapshots --json | jq '.[] | select(.exit_code == 3)'`

That of course returns JSON. If you want to return the snapshot IDs instead, you just make it: `restic snapshots --json | jq -r '.[] | select(.exit_code == 3).id'`
